File: go/core/fargate_metadata_test.go
// Fargate ECS task metadata — single HTTP call per process, cached.
// Decision #7 pin: MiB → bytes via BINARY divisor (1024^3), NOT decimal GB.
//
// Mirrors python/tests/test_fargate_metadata.py.

package core

import (
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"testing"
)

func TestFargateReturnsVCPUAndMemory(t *testing.T) {
	ResetFargateMetadataForTests()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`{"TaskARN":"arn:aws:ecs:us-east-1:0:task/abc","Limits":{"CPU":0.5,"Memory":1024}}`))
	}))
	defer srv.Close()

	t.Setenv("ECS_CONTAINER_METADATA_URI_V4", srv.URL)
	t.Setenv("ECS_CONTAINER_METADATA_URI", "")

	m := FetchFargateMetadata()
	if m == nil {
		t.Fatal("expected metadata, got nil")
	}
	if m.VCPUCount != 0.5 {
		t.Fatalf("VCPUCount = %v, want 0.5", m.VCPUCount)
	}
	// Decision #7 — 1024 MiB → bytes via binary GiB.
	if m.MemoryBytesLimit != int64(1024)*1024*1024 {
		t.Fatalf("MemoryBytesLimit = %d, want %d (Decision #7 binary GiB)",
			m.MemoryBytesLimit, int64(1024)*1024*1024)
	}
}

func TestFargateNoEnvVarReturnsNil(t *testing.T) {
	ResetFargateMetadataForTests()
	t.Setenv("ECS_CONTAINER_METADATA_URI_V4", "")
	t.Setenv("ECS_CONTAINER_METADATA_URI", "")

	if m := FetchFargateMetadata(); m != nil {
		t.Fatalf("expected nil, got %+v", m)
	}
}

func TestFargateUnreachableReturnsNilAndLogsOnce(t *testing.T) {
	ResetFargateMetadataForTests()
	// Use a port that's guaranteed-closed quickly (close immediately).
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "boom", http.StatusInternalServerError)
	}))
	t.Setenv("ECS_CONTAINER_METADATA_URI_V4", srv.URL)
	t.Setenv("ECS_CONTAINER_METADATA_URI", "")
	srv.Close()

	logs := captureFargateLogs(t)

	if m := FetchFargateMetadata(); m != nil {
		t.Fatalf("expected nil, got %+v", m)
	}
	if m := FetchFargateMetadata(); m != nil {
		t.Fatalf("expected nil on second call, got %+v", m)
	}
	if got := atomic.LoadInt32(logs); got != 1 {
		t.Fatalf("log count = %d, want 1 (convention §11 log-once-per-mode)", got)
	}
}

func TestFargateCachedAfterFirstSuccess(t *testing.T) {
	ResetFargateMetadataForTests()
	var hits int64
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		atomic.AddInt64(&hits, 1)
		_, _ = w.Write([]byte(`{"Limits":{"CPU":1,"Memory":512}}`))
	}))
	defer srv.Close()

	t.Setenv("ECS_CONTAINER_METADATA_URI_V4", srv.URL)
	t.Setenv("ECS_CONTAINER_METADATA_URI", "")

	a := FetchFargateMetadata()
	b := FetchFargateMetadata()
	if a == nil || b == nil {
		t.Fatal("expected non-nil")
	}
	if a != b {
		t.Fatal("expected same pointer on cached call")
	}
	if got := atomic.LoadInt64(&hits); got != 1 {
		t.Fatalf("hits = %d, want 1 (cached)", got)
	}
}

func TestFargateMalformedLimitsReturnsNil(t *testing.T) {
	ResetFargateMetadataForTests()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`{"Limits":{"CPU":"garbage"}}`))
	}))
	defer srv.Close()

	t.Setenv("ECS_CONTAINER_METADATA_URI_V4", srv.URL)
	t.Setenv("ECS_CONTAINER_METADATA_URI", "")

	if m := FetchFargateMetadata(); m != nil {
		t.Fatalf("expected nil, got %+v", m)
	}
}

func TestFargateV3URIAlsoWorks(t *testing.T) {
	ResetFargateMetadataForTests()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`{"Limits":{"CPU":2,"Memory":4096}}`))
	}))
	defer srv.Close()

	// Only the v3 (no _V4 suffix) env var set.
	t.Setenv("ECS_CONTAINER_METADATA_URI_V4", "")
	t.Setenv("ECS_CONTAINER_METADATA_URI", srv.URL)

	m := FetchFargateMetadata()
	if m == nil {
		t.Fatal("expected non-nil")
	}
	if m.VCPUCount != 2.0 {
		t.Fatalf("VCPUCount = %v, want 2.0", m.VCPUCount)
	}
	if m.MemoryBytesLimit != int64(4096)*1024*1024 {
		t.Fatalf("MemoryBytesLimit = %d, want %d", m.MemoryBytesLimit, int64(4096)*1024*1024)
	}
}

// TestFargateUsesBinaryGiBDivisor is the load-bearing Decision #7 pin —
// confusing MiB (binary) with MB (decimal) silently over-attributes
// memory cost by ~4.86% on every Fargate task.
func TestFargateUsesBinaryGiBDivisor(t *testing.T) {
	ResetFargateMetadataForTests()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`{"Limits":{"CPU":1,"Memory":1024}}`))
	}))
	defer srv.Close()
	t.Setenv("ECS_CONTAINER_METADATA_URI_V4", srv.URL)
	t.Setenv("ECS_CONTAINER_METADATA_URI", "")

	m := FetchFargateMetadata()
	if m == nil {
		t.Fatal("expected metadata")
	}
	if m.MemoryBytesLimit == 1_073_741_824 {
		// OK: 1024 * 1024 * 1024 — binary GiB.
		return
	}
	if m.MemoryBytesLimit == 1_000_000_000 {
		t.Fatalf("regression: Fargate using DECIMAL GB divisor (1024 MiB → %d bytes) — "+
			"Decision #7 requires BINARY divisor 1024*1024*1024 = 1_073_741_824. "+
			"This is the silent ~4.86%% over-attribution bug.", m.MemoryBytesLimit)
	}
	t.Fatalf("unexpected MemoryBytesLimit = %d", m.MemoryBytesLimit)
}

File: typescript/tests/fargate-metadata.test.ts
/**
 * Fargate ECS task metadata helper.
 *
 * Single HTTP call per process, cached. Exposes vcpuCount (number) and
 * memoryBytesLimit (number — converted from MiB per Decision #7; Fargate
 * uses BINARY MiB, not decimal MB, which is the ~4.86% silent-over-
 * attribution bug the conversion table prevents).
 *
 * Ports python/tests/test_fargate_metadata.py (6 cases) to vitest.
 */

import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import {
  fetchFargateMetadata,
  _resetForTests,
  _setFetchForTests,
} from "../src/core/fargate-metadata.js";

let snapshot: Record<string, string | undefined> = {};

beforeEach(() => {
  _resetForTests();
  snapshot = {
    ECS_CONTAINER_METADATA_URI_V4: process.env.ECS_CONTAINER_METADATA_URI_V4,
    ECS_CONTAINER_METADATA_URI: process.env.ECS_CONTAINER_METADATA_URI,
  };
  delete process.env.ECS_CONTAINER_METADATA_URI_V4;
  delete process.env.ECS_CONTAINER_METADATA_URI;
});

afterEach(() => {
  _setFetchForTests(null);
  _resetForTests();
  for (const [k, v] of Object.entries(snapshot)) {
    if (v === undefined) delete process.env[k];
    else process.env[k] = v;
  }
});

function makeFetch(
  handler: (url: string) => Response | Promise<Response>,
): typeof fetch {
  return ((input: RequestInfo | URL) => {
    const url = typeof input === "string" ? input : input.toString();
    return Promise.resolve(handler(url));
  }) as typeof fetch;
}

describe("fetchFargateMetadata", () => {
  test("returns vcpu and memory (MiB → bytes)", async () => {
    process.env.ECS_CONTAINER_METADATA_URI_V4 = "http://169.254.170.2/v4/abc";
    _setFetchForTests(
      makeFetch(
        () =>
          new Response(
            JSON.stringify({
              TaskARN: "arn:aws:ecs:us-east-1:0:task/abc",
              Limits: { CPU: 0.5, Memory: 1024 },
            }),
          ),
      ),
    );

    const m = await fetchFargateMetadata();
    expect(m).not.toBeNull();
    expect(m!.vcpuCount).toBe(0.5);
    // 1024 MiB → bytes via binary GiB (Decision #7)
    expect(m!.memoryBytesLimit).toBe(1024 * 1024 * 1024);
  });

  test("no env var returns null", async () => {
    expect(await fetchFargateMetadata()).toBeNull();
  });

  test("unreachable returns null and logs once", async () => {
    process.env.ECS_CONTAINER_METADATA_URI_V4 = "http://169.254.170.2/v4/abc";
    const warn = vi.spyOn(console, "warn").mockImplementation(() => {});
    _setFetchForTests(
      makeFetch(() => {
        throw new Error("network unreachable");
      }),
    );

    expect(await fetchFargateMetadata()).toBeNull();
    expect(await fetchFargateMetadata()).toBeNull();

    const fargateLogs = warn.mock.calls.filter((args) =>
      args.some(
        (a) => typeof a === "string" && a.toLowerCase().includes("fargate metadata"),
      ),
    );
    expect(fargateLogs.length).toBe(1);
    warn.mockRestore();
  });

  test("cached after first success", async () => {
    process.env.ECS_CONTAINER_METADATA_URI_V4 = "http://169.254.170.2/v4/abc";
    let calls = 0;
    _setFetchForTests(
      makeFetch(() => {
        calls += 1;
        return new Response(JSON.stringify({ Limits: { CPU: 1, Memory: 512 } }));
      }),
    );

    const a = await fetchFargateMetadata();
    const b = await fetchFargateMetadata();
    expect(a).not.toBeNull();
    expect(a).toBe(b);
    expect(calls).toBe(1);
  });

  test("malformed Limits returns null", async () => {
    process.env.ECS_CONTAINER_METADATA_URI_V4 = "http://169.254.170.2/v4/abc";
    _setFetchForTests(
      makeFetch(
        () =>
          new Response(JSON.stringify({ Limits: { CPU: "garbage" } })),
      ),
    );
    expect(await fetchFargateMetadata()).toBeNull();
  });

  test("v3 URI (ECS_CONTAINER_METADATA_URI, no _V4 suffix) is also valid", async () => {
    process.env.ECS_CONTAINER_METADATA_URI = "http://169.254.170.2/v3/abc";
    _setFetchForTests(
      makeFetch(
        () =>
          new Response(JSON.stringify({ Limits: { CPU: 2, Memory: 4096 } })),
      ),
    );
    const m = await fetchFargateMetadata();
    expect(m).not.toBeNull();
    expect(m!.vcpuCount).toBe(2);
    expect(m!.memoryBytesLimit).toBe(4096 * 1024 * 1024);
  });
});
